### PR TITLE
GPU fixes + remove FrEIA dependency + fix patchNR demo

### PR DESCRIPTION
**Pull request by tachella**
- Updates `pixi` config in `pyproject.toml` to install GPU torch versions, which allows running the CI on GPU (not currently the case)
- Removes `FrEIA` dependency, which breaks the pixi resolve, and is used very little in the library - replace with a custom pytorch implementation of normalizing flow
- Fix `PatchNR` demo, by fixing the trainer, and updating pretrained weights

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [ ] `black .` and `ruff check .` run successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### CODE REVIEW

File: deepinv/optim/prior.py [hunk 1]
```diff
@@ -605,30 +606,209 @@ def fn(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
...
-        When set to ``None`` it is set to a dense invertible neural network built with the FrEIA library, where the number of
-        invertible blocks and the size of the subnetworks is determined by the parameters `num_layers` and `sub_net_size`.
-    :param str pretrained: Define pretrained weights by its path to a `.pt` file, None for random initialization,
-        `"PatchNR_lodopab_small"` for the weights from the limited-angle CT example.
+        supporting backpropagation. It takes a (batched) tensor of flattened patches and the boolean ``rev`` (default ``False``)
+        as input and returns ``(latent, log_det_jacobian)`` as output.
+        If ``rev=True``, it applies the inverse of the flow.
+        When set to ``None``, a GLOW-style invertible neural network is built, where the number of
+        coupling blocks and the hidden neurons of the sub-networks are determined by ``num_layers`` and ``sub_net_size`` respectively.
+        If `None`, it is set to :class:`deepinv.optim.prior.NormalizingFlow` model.
```
Comment (mh-nguyen712) [inline deepinv/optim/prior.py:804]: add `NormalizingFlow` to the rst file

File: deepinv/optim/prior.py [hunk 2]
```diff
@@ -605,30 +606,209 @@ def fn(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
...
+    using :class:`deepinv.optim.prior.GLOWCouplingBlock` as the building block.
+
+    :param int dimension: dimension of each input sample (flattened patch size).
+    :param int num_layers: number of coupling blocks to stack.
+    :param callable subnet: a callable ``subnet(channels_in, channels_out) -> nn.Module``
+        that constructs the subnetworks used inside each coupling block.
+
+        For example
+
+        ::
```
Comment (mh-nguyen712) [inline deepinv/optim/prior.py]: is this the same as .. examples:: ?

File: deepinv/optim/prior.py [hunk 3]
```diff
@@ -605,30 +606,209 @@ def fn(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
...
+    r"""Sequential normalizing flow built from GLOW-style affine coupling blocks.
+
+    The flow maps an input sample ``x`` to a latent representation ``z`` by passing
+    it through ``num_layers`` invertible coupling blocks in sequence.  The
+    log-determinant of the full Jacobian is accumulated additively across blocks.
+    Setting ``rev=True`` runs the blocks in reverse order to recover the original
+    sample from a latent code.
+
+    The architecture follows the generative flow of :footcite:t:`kingma2018glow`,
+    using :class:`deepinv.optim.prior.GLOWCouplingBlock` as the building block.
```
Comment (mh-nguyen712) [inline deepinv/optim/prior.py:708]: add to rst file

File: deepinv/optim/prior.py [hunk 4]
```diff
@@ -605,30 +606,209 @@ def fn(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
...
+    The two-step affine coupling structure follows :footcite:t:`dinh2017density`, and
+    the soft-clamping of scales is introduced in :footcite:t:`kingma2018glow`.
+
+    :param int dim: total input/output dimension (will be split evenly).
+    :param callable subnet: a callable ``subnet(channels_in, channels_out) -> nn.Module``
+        that constructs the subnetworks used inside each coupling step.
+
+        For example:
+
+        ::
```
Comment (mh-nguyen712) [inline deepinv/optim/prior.py]: it's a bit different than the usual example (|sep|, followed by .. examples::)

File: deepinv/optim/prior.py [hunk 5]
```diff
@@ -605,30 +606,209 @@ def fn(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
...
+    r"""GLOW-style affine coupling block.
+
+    Each block performs two successive affine coupling steps on the input vector,
+    which is split into two halves ``(x1, x2)``:
+
+    * Step 1 — a subnetwork acting on ``x1`` produces a pointwise scale ``s1`` and
+      shift ``t1`` that are applied to ``x2``: ``y2 = x2 * exp(s1) + t1``.
+    * Step 2 — a second subnetwork acting on ``y2`` produces ``s2, t2`` that are
+      applied to ``x1``: ``y1 = x1 * exp(s2) + t2``.
+
```
Comment (mh-nguyen712) [inline deepinv/optim/prior.py]: can we write this in math mode?

File: deepinv/optim/prior.py [hunk 6]
```diff
@@ -605,30 +606,209 @@ def fn(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
...
+    Both steps are exactly invertible, and their combined log-determinant of the
+    Jacobian is ``sum(s1) + sum(s2)``.  The scale outputs are soft-clamped via
+    ``clamp * (2/π) * arctan(s / clamp)`` to keep the log-determinant bounded and
```
Comment (mh-nguyen712) [inline deepinv/optim/prior.py]: change the pi into latex

File: pyproject.toml [hunk 7]
```diff
@@ -84,8 +84,8 @@ denoisers = [
     "timm",
     "PyWavelets",
     "ptwt",
-    "FrEIA",
     "kornia",
+     # "FrEIA",  # handled separately in [tool.pixi.pypi-dependencies]
```
Comment (mh-nguyen712) [inline pyproject.toml]: FrEIA is no longer needed, right? if yes, remove it from the dependency list below

**Review by mh-nguyen712 — CHANGES_REQUESTED** [review]
Thanks for the fixes! The docs is not passing due to unknow class ref.

File: deepinv/optim/prior.py (hunk 1 above)
Comment (tachella) [reply deepinv/optim/prior.py:804]: Thanks, done

File: deepinv/optim/prior.py (hunk 2 above)
Comment (tachella) [reply deepinv/optim/prior.py]: I removed and instead added a full doctest example on PatchNR and NormalizingFlow

File: deepinv/optim/prior.py (hunk 3 above)
Comment (tachella) [reply deepinv/optim/prior.py:708]: Done

File: deepinv/optim/prior.py (hunk 5 above)
Comment (tachella) [reply deepinv/optim/prior.py]: Yes good idea-done

File: deepinv/optim/prior.py (hunk 6 above)
Comment (tachella) [reply deepinv/optim/prior.py]: Done

File: pyproject.toml (hunk 7 above)
Comment (tachella) [reply pyproject.toml]: True, I removed it completely from the pyproject

**Comment by tachella** [comment]
alright, all GPU test and docs seem to be working well now

**Review by mh-nguyen712 — APPROVED** [review]
Last comment then looks good to me, thanks! 

Can you also remove the 
```
    pytest.importorskip(
        "FrEIA",
        reason="This test requires FrEIA. It should be "
        "installed with `pip install FrEIA",
    )
``` in test_optim.py?
